### PR TITLE
Support ECP service options in PAOS request

### PR DESCRIPTION
**Pull request by jdennis**
The OASIS specification "SAML V2.0 Enhanced Client or Proxy Profile
Version 2.0" added ECP service options into the HTTP PAOS header. We
previously were not looking for these optional flags. The function
am_validate_paos_header() was rewritten to correctly parse the service
value and the optional URN flags and return a bitmask of the ECP
service flags.

The following flags were added:

  * ECP_SERVICE_OPTION_CHANNEL_BINDING
  * ECP_SERVICE_OPTION_HOLDER_OF_KEY
  * ECP_SERVICE_OPTION_WANT_AUTHN_SIGNED
  * ECP_SERVICE_OPTION_DELEGATION

The flags are stored in the am_req_cfg_rec.ecp_service_options member.

Currently only the want_authn_signed option is acted upon, it sets the
lasso signature hint when generating a AuthnRequest.

Lasso as of 2.5.0 is missing 3 of the URN flag identifier constants, a
patch was submitted to upstream Lasso to include them and was
accepted. As a fallback they are conditionally added to lasso_compat.h
in case the version of lasso this is compiled against does not yet
have the new constants.

A new utility routine, am_strip_whitespace_and_quotes() was added to
support the header parsing.

A new utility routne, am_str_join() was added to support emitting
diagnostic information, it joins an array of strings into a single
string with a separator.

A new utility routine, *am_ecp_service_options_str() was added so log
messages could include the ECP service option flags in their output.

Signed-off-by: John Dennis <jdennis@redhat.com>

### CODE REVIEW

File: auth_mellon_util.c [hunk 1]
```diff
@@ -1507,32 +1507,104 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+    for (end = end > begin ? end-1 : end; end > begin && isspace(*end); end--);
+
+    /* Nuke beginning & ending quotes */
+    if (end > begin && *begin == '"' && *end == '"') {
+        begin++;
+        end--;
+    }
+
+    /* Null terminate */
+    if (end > begin) {
```
Comment (olavmrk) [inline auth_mellon_util.c]: What if this ends up being i 0-byte string? E.g. if the string is `""`? We would like to null-terminate it in that case as well, otherwise we will get the value `"`?

File: auth_mellon_util.c [hunk 2]
```diff
@@ -1507,32 +1507,104 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+    char *begin, *end;
+
+    /* Skip leading whitespace */
+    for (begin = str; *begin && isspace(*begin); begin++);
+
+    /* Find end of string */
+    for (end = begin; *end; end++);
+
+    /* Skip trailing whitespace */
+    for (end = end > begin ? end-1 : end; end > begin && isspace(*end); end--);
```
Comment (olavmrk) [inline auth_mellon_util.c]: This logic is complicated and hard to read. I suggest extracting the test for `end == begin` out of this for loop and doing an early return in that case -- it is a special case where the string consists of only whitespace. E.g.:

```
if (begin == end) {
  return begin; /* The string consisted of only whitespace. */
}
```

File: auth_mellon_util.c [hunk 3]
```diff
@@ -1507,32 +1507,104 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
- *   true if the PAOS header is valid, false otherwise
+ *   true if the PAOS header is valid, false otherwise. If options is non-NULL
+ *   then the set of option flags is returned there.
  *
  */
-#define PAOS_VERSION_TOKEN "ver=\"" LASSO_PAOS_HREF "\""
-#define ECP_SERVICE_TOKEN  "\"" LASSO_ECP_HREF "\""
-bool am_validate_paos_header(request_rec *r, const char *header)
+bool am_validate_paos_header(request_rec *r, const char *header,
+                             ECPServiceOptions *options_return)
```
Comment (olavmrk) [inline auth_mellon_util.c]: I wonder if it may be time to rename the function to something like `am_parse_paos_header`? It is in effect parsing it, and returning false if it is invalid.

File: auth_mellon_util.c [hunk 4]
```diff
@@ -1540,7 +1612,11 @@ bool am_validate_paos_header(request_rec *r, const char *header)
         goto cleanup;
     }
 
-    /* Split the header on the semicolon character */
+    /*
+     * Split the header on the semicolon character.
+     * There should be 2 items, the first is the version delclaration
+     * and the second is the service and it's options.
+     */
     semicolon_tokens = g_strsplit(header, ";", 0);
```
Comment (olavmrk) [inline auth_mellon_util.c]: Considering the complexity of this header now, I am wondering how safe/unsafe this method of parsing it is. Before it was very strict in what it accepted, but now it tries to deal with all sorts of extra flags/options, but it still uses `g_strsplit()` for parsing, both here, and now also `service_tokens` further down.

Some reasssuring words about how commas and semicolons can never occur as part of neither the version field nor the options would be appreciated. :smiley:

Just to be clear, I am thinking of headers like:

```
PAOS: ver="urn:liberty:paos:2003-08"; "somerandomoption;param=42"
```

I have no idea if that sort of thing is supposed to be possible, but the fact that they allow (or encourage?) the use of double-quotes makes me wonder about it.

File: auth_mellon_util.c [hunk 5]
```diff
@@ -1827,7 +1947,9 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
     }
 
     if (selected_descriptor) {
-        url = lasso_provider_get_metadata_one(provider, selected_descriptor);
+        url = lasso_provider_get_metadata_one_for_role(provider,
+                                                       LASSO_PROVIDER_ROLE_SP,
+                                                       selected_descriptor);
```
Comment (olavmrk) [inline auth_mellon_util.c]: This change is a bit surprising -- does it belong in a separate patch?

File: auth_mellon_util.c [hunk 6]
```diff
@@ -1835,7 +1957,130 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
...
+    apr_size_t alloc_len;
+    apr_size_t n_strings = 0;
+    apr_size_t total_string_len = 0;
+    apr_size_t separator_len = strlen(separator);
+
+    /*
+     * Iterate over string array, compute total number of characters
+     * and a count of how many strings
+     */
+    for (total_string_len = 0, n_strings = 0, p = strings; *p; p++) {
```
Comment (olavmrk) [inline auth_mellon_util.c]: I think this loop would be much clearer if the initialization of `total_string_len` and `n_strings` happened outside of the for-loop. Including them in the for-loop like this seems to imply that it is something it iterates over, rather than what it actually counts.

File: auth_mellon_util.c [hunk 7]
```diff
@@ -1835,7 +1957,130 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
...
+        total_string_len += strlen(*p);
+        n_strings++;
+    }
+
+    /*
+     * The allocation length is the length of all the strings
+     * plus the length of all the separators plus 1 for null terminator
+     */
+    alloc_len = total_string_len +
+        (n_strings > 0 ? separator_len*(n_strings-1) : 0) + 1;
```
Comment (olavmrk) [inline auth_mellon_util.c]: If you dropped the `alloc_len` variable, you could change this to something like:

```
if (n_strings > 0) {
  total_string_len += separator_len * (n_strings - 1);
}
```

Which I think is far more readable.

File: auth_mellon_util.c [hunk 8]
```diff
@@ -1835,7 +1957,130 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
...
+
+    str = apr_palloc(pool, alloc_len);
```
Comment (olavmrk) [inline auth_mellon_util.c]: Would in that case of course have to change this to be `apr_palloc(pool, total_string_len + 1)`.

File: auth_mellon_util.c [hunk 9]
```diff
@@ -1835,7 +1957,130 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
...
+ * extra string will be appended of the form "(unknown bits = x)".
+ *
+ * Parameters:
+ *  pool    memory allocation pool
+ *  options bitmask of PAOS options
+ */
+char *am_ecp_service_options_str(apr_pool_t *pool, ECPServiceOptions options)
+{
+    int i = 0;
+    char *names[10];            /* Warning, increase this if flags are added,
```
Comment (olavmrk) [inline auth_mellon_util.c]: Doesn't this need to be `const char *` to avoid some warnings, since you are assigning constant strings to it?

**Comment by olavmrk** [comment]
Hi, I started looking through this patch today, but I have to leave now.

One thing that I am wondering about is the unhandled options. Is it safe to ignore unknown and unhandled ECP options, or should they raise some sort of error, for future-compatibility?

File: auth_mellon_util.c (hunk 3 above)
Comment (jdennis) [inline auth_mellon_util.c]: I'm fine with changing the name from "validate" to "parse", perfectly reasonable.

File: auth_mellon_util.c (hunk 5 above)
Comment (jdennis) [inline auth_mellon_util.c]: good catch, yes it does.

File: auth_mellon_util.c (hunk 6 above)
Comment (jdennis) [inline auth_mellon_util.c]: sure, those variables are initialized in their declaration so the initialization in the loop start is redundant anyway, was just trying to be explicit.

File: auth_mellon_util.c (hunk 7 above)
Comment (jdennis) [inline auth_mellon_util.c]: sounds reasonable

File: auth_mellon_util.c (hunk 9 above)
Comment (jdennis) [inline auth_mellon_util.c]: done

File: auth_mellon_util.c (hunk 4 above)
Comment (jdennis) [inline auth_mellon_util.c]: You have a good point and I wondered about this myself. I did look at all the URN's used in SAML and none had commas or semicolons in them. As far as I know only URN's  will ever appear in this header. However the fact the URN's are quoted which according to the HTTP parsing rules turns them into a string token suggests the committee was forward looking to protect against embedded token separators inside a URN string token.

So I thought about the problem. Should I do real parsing, break the header into individual tokens and then operate on the sequence of tokens (maybe not as general as building a parse tree). My thought was, that is is awful complicated, especially just for one HTTP header. Writing robust parsing code in C is non-trivial (look I even made mistake in the  am_strip_whitespace_and_quotes function which you caught). I know we do a better job of parsing the headers than Shibboleth does because our code caught a mistake in both the SAML spec and in Shibbolth in the Accept header. The spec incorrectly used a semicolon separator when it should have been a comma and Shibboleth never caught because all they do is a string search for a substring and ignore everything else in the header. I've since gotten OASIS to fix the spec and filed a bug with Shibboleth.

My guess is this code is probably still more robust than Shibboleth, but you're right it's imperfect in the service of simplicity. I don't think we'll see URN's with embedded comma's or semicolons, none are currently defined, but of course there is no guarantee one won't be defined in the future.

I might take a crack at parsing the header into tokens and then doing a simple analysis of the grammar. This would also eliminate the need for the am_strip_whitespace_and_quotes() function which isn't 100% correct either because it accepts non-quoted strings and does not handle embedded quote characters which was another simplification I was willing to accept in the service of simplicity.

**Comment by jdennis** [comment]
I added comments in the code in response to your review.

As for the unhandled options I think we should emit a warning in the log for the time being. I'm not aware of any software at the moment which is utilizing them, logging a warning seems sufficient for the moment and support can be added if and when the need arises.

File: auth_mellon_util.c (hunk 4 above)
Comment (olavmrk) [inline auth_mellon_util.c]: It might be worth checking if Apache has some utility routines for parsing this kind of "maybe-quoted" header strings that can be used by modules.

If one ends up writing a full parser from scratch, I would guess it isn't worth it at the moment. It would be nice if one somehow could ensure that the entire thing doesn't blow up when data that isn't handled by the parser shows up though.

**Comment by olavmrk** [comment]
I'm mostly worried that someone may set a flag that we don't handle at the moment, and then if support is added for that flag later, it would suddenly break SPs that set a flag that we ignored. I don't know if that is a realistic scenario in this case though. I also don't know what the spec says about processing of unknown options. I.e. should they be ignored, or should an error be returned.

File: auth_mellon_handler.c [hunk 10]
```diff
@@ -2761,6 +2761,39 @@ static int am_init_authn_request_common(request_rec *r,
...
+        } else {
+            /*
+             * The client did not explicitly the request be signed. To
+             * be compatible with earlier versions of ECP which did
+             * not have this flag default to respecting the metadata
+             * and/or the configuration, the lasso
+             * LASSO_PROFILE_SIGNATURE_HINT_MAYBE covers this case.
+             */
+            lasso_profile_set_signature_hint(LASSO_PROFILE(login),
+                                             LASSO_PROFILE_SIGNATURE_HINT_MAYBE);
```
Comment (olavmrk) [inline auth_mellon_handler.c]: This behavior is something I wonder slightly about -- why do we need to set `LASSO_PROFILE_SIGNATURE_HINT_MAYBE` here? Should that be set unconditionally, even if `HAVE_ECP` isn't set?

File: auth_mellon_util.c (hunk 4 above)
Comment (jdennis) [inline auth_mellon_util.c]: Back when I first started working on this I checked Apache to see if there was utility functions I could call to parse the header, there wasn't. But I checked again to be sure. Sadly there is no common code, it appears as if every module implements their own hand crafted code. I suppose that shouldn't be surprising because HTTP header fields are quirky and non-uniform, every header seems to be just a little bit different.

But anyway I rewrote the parser to tokenize and then analyze the token sequence. It has very good diagnostics and will report the exact location in the string where it detected a problem. I tested the parser in stand-alone mode it handled anything I threw at it.

File: auth_mellon_handler.c (hunk 10 above)
Comment (jdennis) [inline auth_mellon_handler.c]: This is logically equivalent to unconditionally setting it to LASSO_PROFILE_SIGNATURE_HINT_MAYBE for the ECP case. The default in Lasso is LASSO_PROFILE_SIGNATURE_HINT_MAYBE, as a consequence I've removed this block, the default should be in effect.

**Comment by jdennis** [comment]
On 10/29/2015 07:49 AM, Olav Morken wrote:
>
> I'm mostly worried that someone may set a flag that we don't handle at 
> the moment, and then if support is added for that flag later, it would 
> suddenly break SPs that set a flag that we ignored. I don't know if 
> that is a realistic scenario in this case though. I also don't know 
> what the spec says about processing of unknown options. I.e. should 
> they be ignored, or should an error be returned.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/UNINETT/mod_auth_mellon/pull/55#issuecomment-152155862>.
>
I added logic which detects if any unsupported options are passed, if so 
we return a 501 Not Implemented error.

**Comment by jdennis** [comment]
branch was updated with code fixes discussed in review

**Comment by olavmrk** [comment]
Hi, just want to let you know that I unfortunately didn't have the time to look at this today, and I am traveling tomorrow. I hope to be able to take a look at it on Wednesday.

File: auth_mellon_util.c [hunk 11]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+#define PUSH_TOKEN(token_type, start, end)              \
+{                                                       \
+    const char *s = (start);                            \
+    const char *e = (end);                              \
+    apr_size_t offset = s - str;                        \
+                                                        \
+    if (token_type == TOKEN_DBL_QUOTE_STRING) {         \
+        /* do not include quotes in token value */      \
+        s++; e--;                                       \
+    }                                                   \
```
Comment (olavmrk) [inline auth_mellon_util.c]: I think you also need to convert any cases of `\"` and similar into `"` when processing double quoted strings.

File: auth_mellon_util.c [hunk 12]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+                                       underscore */
+            p++;
+            while(*p && (apr_isalnum(*p) || *p == '_')) p++;
+            PUSH_TOKEN(TOKEN_IDENTIFIER, start, p);
+            start = p;
+        }
+        else if (*p == '"') {   /* double quoted string,
+                                   quote escaped with backslash */
+            p++;
+            while(*p && (*p != '\"' || *(p-1) == '\\')) p++;
```
Comment (olavmrk) [inline auth_mellon_util.c]: I think you must have a slightly more complicated loop here, to deal with `"\\"`.

File: auth_mellon_util.c [hunk 13]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+ *  str               input string to be parsed.
+ *  ignore_whitespace if True whitespace tokens are not returned
+ *  error             location where error string is returned
+ *                    if NULL no error occurred
+ * Returns:           array of Token objects
+ */
+
+#define TOKEN_INIT(t) memset(&t, 0, sizeof(t))
+
+#define PUSH_TOKEN(token_type, start, end)              \
```
Comment (olavmrk) [inline auth_mellon_util.c]: Is one allowed to access `tokens->pool`? In that case, I would suggest replacing this `#define` with a `static inline` function:

```
static inline void push_token(apr_array_header_t *tokens, token_type, start, end)
```

Using a `#define` makes the code hard to read, especially as you try to match variables.

File: auth_mellon_util.c [hunk 14]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+        s++; e--;                                       \
+    }                                                   \
+                                                        \
+    token.type = token_type;                            \
+    token.len = e-s;                                    \
+    token.offset = offset;                              \
+    token.str = apr_pstrmemdup(pool, s, token.len);     \
+    APR_ARRAY_PUSH(tokens, Token) = token;              \
+                                                        \
+    TOKEN_INIT(token);                                  \
```
Comment (olavmrk) [inline auth_mellon_util.c]: Instead of using a single token variable across all instances of this `#define`, couldn't you simply change the code to something like:

```
Token *token = apr_array_push(tokens);
token->type = token_type;
[...]
```

File: auth_mellon_util.c [hunk 15]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
-    /* Split the header on the semicolon character */
-    semicolon_tokens = g_strsplit(header, ";", 0);
-
-    /* There must be exactly two tokens after splitting */
-    len = g_strv_length(semicolon_tokens);
-    if (len != 2) {
+    /* Header must begin with "ver=xxx" where xxx is paos version */
+    if (!is_token(tokens, 0, TOKEN_IDENTIFIER, "ver") ||
+        !is_token(tokens, 1, TOKEN_EQUAL, NULL) ||
+        !is_token(tokens, 2, TOKEN_DBL_QUOTE_STRING, LASSO_PAOS_HREF)) {
```
Comment (olavmrk) [inline auth_mellon_util.c]: Are the double-quotes actually required to be part of these strings? Or are you allowed to drop them? I'm just curious if one should have a test for "double-quoted or identifier"?

File: auth_mellon_util.c [hunk 16]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+ *   urn:oasis:names:tc:SAML:2.0:cm:holder-of-key
+ *
+ * Request for signed SAML request
+ *   urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp:2.0:WantAuthnRequestsSigned
+ *
+ * Request to delegate credentials to the service provider
+ *   urn:oasis:names:tc:SAML:2.0:conditions:delegation
+ *
+ *
+ * Example PAOS HTP header::
```
Comment (olavmrk) [inline auth_mellon_util.c]: Missing "T" in HTTP, and double colon here.

File: auth_mellon_util.c [hunk 17]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+    /* Next is the service value, separated from the version by a semicolon */
+    if (!is_token(tokens, 3, TOKEN_SEMICOLON, NULL)) {
         ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
                      "invalid PAOS header, "
-                     "expected first token to be \"%s\", "
-                     "but found \"%s\" in header=\"%s\"",
-                     PAOS_VERSION_TOKEN, token, header);
+                     "expected semicolon after PAOS version "
+                     "but found \"%s\" at position %lu in header=\"%s\"",
+                      APR_ARRAY_IDX(tokens, 3, Token).str,
```
Comment (olavmrk) [inline auth_mellon_util.c]: Here you read beyond the end of the array if the tokenized string ends after the double-quoted version. I.e.:
```
PAOS: ver="urn:liberty:paos:2003-08"
```

File: auth_mellon_util.c [hunk 18]
```diff
@@ -1835,7 +2147,127 @@ char *am_get_assertion_consumer_service_by_binding(LassoProvider *provider, cons
...
+        names[i++] = "delegation";
+        options &= ~ECP_SERVICE_OPTION_DELEGATION;
+    }
+
+    if (options) {
+        names[i++] = apr_psprintf(pool, "(unknown bits = %#x)", options);
+    }
+
+    names[i] = NULL;
+    return am_str_join(pool, names, ", ");
```
Comment (olavmrk) [inline auth_mellon_util.c]: After realizing that there is a [`apr_array_pstrcat`](https://apr.apache.org/docs/apr/2.0/group__apr__tables.html#ga6232817e4a32ee5619c99fe55e145c7a)-function in APR, I think this function could be rewritten using that function. I think that would be better than rolling our own version of that function. It would mean that the elements would only be separated by `','`, but I don't think that is a problem.

File: auth_mellon_util.c [hunk 19]
```diff
@@ -1852,7 +2285,7 @@ bool am_is_paos_request(request_rec *r)
         }
     }
     if (paos_header) {
-        if (am_validate_paos_header(r, paos_header)) {
+        if (am_parse_paos_header(r, paos_header, &ecp_service_options)) {
             valid_paos_header = true;
```
Comment (olavmrk) [inline auth_mellon_util.c]: Should a parse error in the PAOS-header trigger a "400 Bad Request"?

File: auth_mellon_util.c [hunk 20]
```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
         ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
-                     "invalid PAOS header, "
-                     "expected 2 tokens seperated by semicolon, header=\"%s\"",
-                     header);
+                      "invalid PAOS header, "
+                      "expected header to begin with ver=\"%s\", "
+                      "actual header=\"%s\"",
+                      LASSO_PAOS_HREF, header);
         goto cleanup;
+
```
Comment (olavmrk) [inline auth_mellon_util.c]: This stray line feed looks out of place.

File: auth_mellon_util.c (hunk 11 above)
Comment (jdennis) [inline auth_mellon_util.c]: Fixed.  Figuring out what the exact rules are with respect to escapes was a bit of an effort but the comments in the code clearly spell it out.

File: auth_mellon_util.c (hunk 12 above)
Comment (jdennis) [inline auth_mellon_util.c]: Yup, thanks. When I wrote that it didnt' feel right. Fixed now.

File: auth_mellon_util.c (hunk 13 above)
Comment (jdennis) [inline auth_mellon_util.c]: Fair enough, it's an inline function now. Fixed.

File: auth_mellon_util.c (hunk 14 above)
Comment (jdennis) [inline auth_mellon_util.c]: Good suggestion. Fixed.

File: auth_mellon_util.c (hunk 15 above)
Comment (jdennis) [inline auth_mellon_util.c]: As I read the spec, yes the double quotes are required. Code will remain as is.

File: auth_mellon_util.c (hunk 16 above)
Comment (jdennis) [inline auth_mellon_util.c]: Fixed.

File: auth_mellon_util.c (hunk 17 above)
Comment (jdennis) [inline auth_mellon_util.c]: Good idea, done.

File: auth_mellon_util.c (hunk 18 above)
Comment (jdennis) [inline auth_mellon_util.c]: Good suggestion. Done

File: auth_mellon_util.c (hunk 20 above)
Comment (jdennis) [inline auth_mellon_util.c]: Fixed.

File: auth_mellon_util.c (hunk 19 above)
Comment (jdennis) [inline auth_mellon_util.c]: Yes, returning bad request would be the right behavior. Added error code return to am_parse_paos_header() and check it.

**Comment by jdennis** [comment]
Patch has been reworked with suggested fixes and is ready for re-review.

File: auth_mellon_util.c [hunk 21]
```diff
@@ -1507,83 +1507,474 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
...
+                  "PAOS header: \"%s\"", header);
+
+    tokens = tokenize(r->pool, header, true, &error);
+
+#ifdef DEBUG
+    dump_tokens(r, tokens);
+#endif
+
+    if (error) {
+        ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r, error);
```
Comment (olavmrk) [inline auth_mellon_util.c]: Hi,

sorry for not picking up on this earlier, but this line triggers a warning:

```
auth_mellon_util.c:1872:9: warning: format not a string literal and no format arguments [-Wformat-security]
```

Looking at it more closely, I think this would be a vulnerability, since `error` can contain the actual header that was parsed.

I guess something like this should be a simple solution:

```
ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r, "%s", error);
```

Comment (jdennis) [inline auth_mellon_util.c]: yes, that's the right fix and I should have caught it as well. It's fixed now and branch is updated.

**Comment by olavmrk** [comment]
Thanks! I have merged your pull request now.
